programs/survival/holy_hand_grenades.sc

```diff
@@ -50,16 +51,14 @@ __create_bullet(player, power) ->
 );
 
 __explode(entity, direction_vec, power) -> (
-	l(x,y,z) = pos(entity);
-	block = first(diamond(x,y,z,3,3), !__is_invalid_for_smashing(_));
-	if (block, __cascade_smash(pos(entity), 1.2*power, direction_vec, pos(block)))
+	block = first(diamond(pos(entity),3,3), !__is_invalid_for_smashing(_));
+	if (block, __cascade_smash(pos(entity), 1.2*power, direction_vec/hardness(pos(entity)), pos(block)))
 );
 
-__is_invalid_for_smashing(block) -> 
+__is_invalid_for_smashing(block) -> (
 	air(block) || block == 'lava' || block == 'water' || 
-	block == 'bedrock' || block == 'barrier' || 
-	block ~ 'command_block' || block == 'structure_block' || block == 'jigsaw';
-
+	block == 'bedrock' || item_category(block)==null;
```

Comment:
Did I miss an item category?

Reply:
Not really, just this is faster, and if more things are added in the future to this list, then this will cover them all. Also it's a good application for people who wanna see how it works.